File: day20/solution.py
#!/usr/bin/env python3
import sys
from functools import cache
from collections import deque

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'example.txt'
D4 = [(-1, 0), (0, 1), (1, 0), (0, -1)]

for arg in sys.argv[1:]:
  if arg == '-i': FILENAME = 'input.txt'
  elif arg == '-e': FILENAME = 'example.txt'
  elif arg.startswith('-e'): FILENAME = f'example{arg[-1]}.txt'
  elif arg == '-d': DEBUG = 1
  elif arg == '-d2': DEBUG = 2
  elif arg == '-d3': DEBUG = 3
  else: raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  text = f.read()
# if DEBUG: print(f'{text=}')
map_data = [list(x) for x in text.splitlines()]
rows = len(map_data)
cols = len(map_data[0])
start_coord = end_coord = None
for r in range(rows):
  for c in range(cols):
    match map_data[r][c]:
      case 'S':
        start_coord = (r, c)
      case 'E':
        end_coord = (r, c)
assert start_coord is not None and end_coord is not None


def draw_map(map_data: list[list[str]]):
  for r in range(rows):
    print(f'''{r: 3d}  {''.join(map_data[r])}''')
  print(f'{rows=} {cols=}')


def get_distances():
  distances = [[-1] * cols for _ in range(rows)]
  cr, cc = start_coord
  distances[cr][cc] = 0
  while (cr, cc) != end_coord:
    if DEBUG > 1: print(f'get_distances {cr=} {cc=}')
    for d in D4:
      nr, nc = cr + d[0], cc + d[1]
      if not (0 <= nr < rows and 0 <= nc < cols): continue
      if distances[nr][nc] != -1: continue
      if map_data[nr][nc] == '#': continue
      distances[nr][nc] = distances[cr][cc] + 1
      cr, cc = nr, nc
      break
  return distances


def part1():
  distances = get_distances()
  # for debugging
  shortcut_time_counts: dict[int, int] = {}
  shortcut_count = 0

  for r in range(rows):
    for c in range(cols):
      origin_distance = distances[r][c]
      if origin_distance == -1: continue
      for d in [
        (-2, 0),
        (-1, 1),
        (0, 2),
        (1, 1),
        (2, 0),
        (1, -1),
        (0, -2),
        (-1, -1),
      ]:
        nr, nc = r + d[0], c + d[1]
        if not (0 <= nr < rows and 0 <= nc < cols): continue
        distance = distances[nr][nc]
        if distance == -1: continue
        saving = distance - origin_distance - 2
        if saving <= 0: continue
        if saving not in shortcut_time_counts.keys():
          shortcut_time_counts[saving] = 1
        else:
          shortcut_time_counts[saving] += 1
        if saving < 100: continue
        shortcut_count += 1

  if DEBUG > 1:
    for k in sorted(shortcut_time_counts.keys()):
      print(f'shortcut count: {k=} v={shortcut_time_counts[k]}')
  print(f'part 1: {shortcut_count=}')


# draw_map(map_data)
part1()

File: day20/solution.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Coord struct {
	r, c int
}
type Cheat struct {
	r, c, t int
}

func main() {
	// d4:=[][]int{{-1, 0},{0, 1},{1, 0},{0, -1}}
	filename, debug := parseArgs()
	fmt.Printf("filename: %s; debug: %d\n", filename, debug)
	mapData, start, end := readData(filename)
	if debug > 0 {
		fmt.Println(mapData)
		fmt.Println("start", start, "end", end)
	}
	part1 := solve(mapData, start, end, 2, debug)
	fmt.Println("part 1:", part1)
	part2 := solve(mapData, start, end, 20, debug)
	fmt.Println("part 2:", part2)
}

func solve(mapData [][]string, start Coord, end Coord, radius int, debug uint8) int {
	rows := len(mapData)
	cols := len(mapData[0])

	// initialise mapTimes
	mapTimes := make([][]int, 0, rows)
	for row := 0; row < rows; row++ {
		row := make([]int, cols)
		for col := 0; col < cols; col++ {
			row[col] = -1
		}
		mapTimes = append(mapTimes, row)
	}

	// walk mapData and populate mapTimes
	current := Coord{start.r, start.c}
	mapTimes[current.r][current.c] = 0
	for current != end {
		for d := 0; d < 4; d++ {
			test := Coord{current.r, current.c}
			switch d {
			case 0:
				test.r--
			case 1:
				test.c++
			case 2:
				test.r++
			case 3:
				test.c--
			}
			// continue on wall or walked (no need for oob check since the map is surrounded by wall)
			if mapData[test.r][test.c] == "#" || mapTimes[test.r][test.c] != -1 {
				continue
			}
			// path is valid, move and break from the direction loop
			mapTimes[test.r][test.c] = mapTimes[current.r][current.c] + 1
			current = test
			break
		}
	}
	if debug > 1 {
		fmt.Println("mapTimes:")
		for _, row := range mapTimes {
			fmt.Println(row)
		}
	}

	// precalculate the available cheat moves
	cheats := []Cheat{}
	for row := 0 - radius; row <= radius; row++ {
		for col := 0 - radius; col <= radius; col++ {
			moves := intAbs(row) + intAbs(col)
			if moves >= 2 && moves <= radius {
				cheats = append(cheats, Cheat{row, col, moves})
			}
		}
	}
	if debug > 1 {
		fmt.Println("cheats:")
		for _, offset := range cheats {
			fmt.Println(offset)
		}
	}

	// test each possible cheat
	count := 0
	for row := 0; row < rows; row++ {
		for col := 0; col < cols; col++ {
			originTime := mapTimes[row][col]
			if originTime == -1 {
				continue
			}
			for _, offset := range cheats {
				nr := row + offset.r
				nc := col + offset.c
				// oob
				if nr < 0 || nr >= rows || nc < 0 || nc >= cols {
					continue
				}
				destTime := mapTimes[nr][nc]
				// wall
				if destTime == -1 {
					continue
				}
				saving := destTime - originTime - offset.t
				if saving < 100 {
					continue
				}
				count++
			}

		}
	}
	return count
}

func intAbs(value int) int {
	if value < 0 {
		return 0 - value
	} else {
		return value
	}
}

func parseArgs() (filename string, debug uint8) {
	filename = "example.txt"
	debug = 0
	for _, arg := range os.Args[1:] {
		switch arg {
		case "-i":
			filename = "input.txt"
		case "-e":
			filename = "example.txt"
		case "-e2":
			filename = "example2.txt"
		case "-e3":
			filename = "example3.txt"
		case "-e4":
			filename = "example4.txt"
		case "-d":
			debug = 1
		case "-d2":
			debug = 2
		case "-d3":
			debug = 3
		default:
			fmt.Printf("unknown arg: %s\n", arg)
		}
	}
	return
}

func readData(filename string) (mapData [][]string, start Coord, end Coord) {
	f, _ := os.Open(filename)
	scanner := bufio.NewScanner(f)
	rowIx := 0
	for scanner.Scan() {
		chars := strings.Split(scanner.Text(), "")
		mapData = append(mapData, chars)
		for colIx, char := range chars {
			switch char {
			case "S":
				start = Coord{rowIx, colIx}
			case "E":
				end = Coord{rowIx, colIx}
			}
		}
		rowIx++
	}
	return
}
